### ISSUE TITLE
Expose parameter annotations 

### ISSUE THREAD

**Issue by wolfs** [TNG/ArchUnit#113]
For defining assertions that method parameters have certain annotations it would be great to expose method parameter annotations via some method on `JavaMethod`.

Currently, I need to do the following to get the annotations of the first parameter:
```
method.reflect().getParameterAnnotations()[0])
```

I used this to check if Java setters have matching `@Nullable` annotations with Java getters here: https://github.com/gradle/gradle/blob/master/subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java#L254-L256

**Comment by wengertj** [comment]
@wolfs I might be wrong (or missunderstand your question) but I think this is already possible.
Looking at the [Javadoc for JavaMember](https://static.javadoc.io/com.tngtech.archunit/archunit/0.9.3/com/tngtech/archunit/core/domain/JavaMember.html) there is a method `getAnnotations()`.

So you should be able to do this:
```java
// Returns com.tngtech.archunit.core.domain.JavaAnnotation
for (final JavaAnnotation annotation : method.getAnnotations()) {
     // Do stuff with:
      annotation.getType();
      annotation.getProperties()
}
```

**Comment by codecholeric** [comment]
I think the problem is 

```
@Foo
void method() { ... }
```

which is already implemented, vs.

```
void method(@Foo param) { ... }
```

which is unfortunately still missing... One bad thing is that there is a `getParameters()` method that returns an iterable of `JavaClass` while we would need sth. like `CodeUnitParameter` that offers `getType()` and `getAnnotations()` (or similar). So this will probably need some migration via `@Deprecated` to get there.

**Comment by wengertj** [comment]
Oh ok I see - I really did not understand the question then :)

**Comment by Farbauti89** [comment]
Hi everyone,

I'm interested in solving this issue.
At the "Hackergarten Berlin" [Philip Riecks](https://github.com/rieckpil) and I already started to work on it :)

**Comment by remal** [comment]
Are there any news?

**Comment by steftux** [comment]
We had another crack at it, unfortunately currently stuck at 
        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
            return new AnnotationProcessor(addAnnotationTo(annotations), annotationBuilderFor(desc));
        }
from MethodProcessor class from JavaClassProcessor.java somehow...

**Comment by Farbauti89** [comment]
Yeah basically we are struggling with the integration of the asm Dependency right now.
The Method we need to extend seems to be [JavaClassProcessor::visitMethod().](https://github.com/TNG/ArchUnit/blob/master/archunit/src/main/java/com/tngtech/archunit/core/importer/JavaClassProcessor.java#L217-L237)

The current problem is that we couldn't figure out how to get the Annotations of the Parameters from asm. I'am pretty sure it is possible with [MethodVisitor](https://asm.ow2.io/javadoc/index.html?org/objectweb/asm/MethodVisitor.html)::visitParameterAnnotation() but yesterday we didn't understand how. To be honest we didn't understand all these "visit"-Methods at all. I guess we need to dig a little bit deeper into asm itself to understand how it works.

However once we get this information it should (at least I hope so) be pretty easy to extend the ArchUnit internals to hold the information. This is the part already prototyped in the Commit from last year.

**Comment by remal** [comment]
@Farbauti89 Yes, you're right, you have to use `MethodVisitor.visitParameterAnnotation`. Basically, you can take a look at `MethodNode` class, which extends `MethodVisitor`.

**Comment by HonoluluHenk** [comment]
is this still in progress?

**Comment by wimdeblauwe** [comment]
Would also need this. In my case to check if there is an `@Qualifier` annotation on a constructor parameter.

**Comment by sergeymeshkov** [comment]
I have the same issue. For example I would like to check dependency between packages
`TestOne` is failed, however `TestTwo` is passed.

```
import javax.validation.constraints.NotBlank;

public class TestOne {
    public void testOne(@NotBlank String one) {
    }
}
```

```
import javax.validation.constraints.NotBlank;

public class TestTwo {
    @NotBlank
    private String two;
}
```


```
    @ArchTest
    public static void validateImport(JavaClasses classes) {
        final var rule = classes()
                .that()
                .resideInAPackage("..test..")
                .should()
                .dependOnClassesThat()
                .resideInAPackage("javax.validation.constraints");

        rule.check(classes);
    }
```

**Comment by codecholeric** [comment]
@Farbauti89 I suppose you guys gave up on this one, huh? :wink: I for my part can only say that it is on my list, if it still isn't solved once I'm done with generics. But for now I'm trying to finish full generics support. Once that is done I will tackle this if it's still open and method references (at least that's my plan at the moment, because I consider that the most pressing core issues)...
PS: I'm about to introduce generic method parameters and I'll likely add those as wrapper objects and not `List<JavaType>` so it can be easily extended with `parameter.getAnnotations()`.
